// @ohoareau/react-document-renderer/lib/blocks/layouts/Row010LayoutBlock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const Block_1 = __importDefault(require("../../Block"));
const renderer_1 = require("@react-pdf/renderer");
const hocs_1 = require("../../hocs");
exports.Row010LayoutBlock = hocs_1.pdfComponent({
    layouts_row010: {
        display: 'flex',
        flexDirection: 'row',
    },
    layouts_row010_left: {
        width: '20%',
        justifyContent: 'flex-start',
    },
    layouts_row010_center: {
        flex: 1,
        justifyContent: 'center',
    },
    layouts_row010_right: {
        width: '20%',
        justifyContent: 'flex-end',
    }
}, ({ s = () => { }, block }) => {
    const content = block.content || {};
    return (react_1.default.createElement(renderer_1.View, { style: s('layouts_row010') },
        !!content.left && react_1.default.createElement(renderer_1.View, { style: s('layouts_row010_left') },
            react_1.default.createElement(Block_1.default, { id: content.left.id, block: content.left })),
        !!content.center && react_1.default.createElement(renderer_1.View, { style: s('layouts_row010_center') },
            react_1.default.createElement(Block_1.default, { id: content.center.id, block: content.center || {} })),
        !!content.right && react_1.default.createElement(renderer_1.View, { style: s('layouts_row010_right') },
            react_1.default.createElement(Block_1.default, { id: content.right.id, block: content.right }))));
});
exports.default = exports.Row010LayoutBlock;
