// @ohoareau/react-document-renderer/lib/blocks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ContentBlock_1 = require("./ContentBlock");
exports.content = ContentBlock_1.default;
var FooterBlock_1 = require("./FooterBlock");
exports.footer = FooterBlock_1.default;
var HeaderBlock_1 = require("./HeaderBlock");
exports.header = HeaderBlock_1.default;
var HeroBlock_1 = require("./HeroBlock");
exports.hero = HeroBlock_1.default;
var ImageBlock_1 = require("./ImageBlock");
exports.image = ImageBlock_1.default;
var LayoutBlock_1 = require("./LayoutBlock");
exports.layout = LayoutBlock_1.default;
var ParagraphBlock_1 = require("./ParagraphBlock");
exports.paragraph = ParagraphBlock_1.default;
var HeadingBlock_1 = require("./HeadingBlock");
exports.heading = HeadingBlock_1.default;
var ListBlock_1 = require("./ListBlock");
exports.list = ListBlock_1.default;
var UnknownBlock_1 = require("./UnknownBlock");
exports.unknown = UnknownBlock_1.default;
var TableBlock_1 = require("./TableBlock");
exports.table = TableBlock_1.default;
var SpaceBlock_1 = require("./SpaceBlock");
exports.space = SpaceBlock_1.default;
var QrCodeBlock_1 = require("./QrCodeBlock");
exports.qrcode = QrCodeBlock_1.default;
var GroupBlock_1 = require("./GroupBlock");
exports.group = GroupBlock_1.default;
