// @ohoareau/lambda-utils/lib/utils/createRouterHandler.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helpers = __importStar(require("../responses"));
const getLambdaConfig_1 = __importDefault(require("./getLambdaConfig"));
const getRequestInfosFromEvent_1 = __importDefault(require("./getRequestInfosFromEvent"));
const mergeConfig_1 = __importDefault(require("./mergeConfig"));
const defaultProcessHandler_1 = __importDefault(require("./defaultProcessHandler"));
function createRouterHandler(defaultConfig = {}, customProcessHandler = undefined) {
    return async function (event, context) {
        var _a, _b, _c;
        try {
            const config = mergeConfig_1.default(await getLambdaConfig_1.default(), defaultConfig);
            const request = getRequestInfosFromEvent_1.default(event, config['routes']);
            const ctx = {
                request,
                config,
                helpers,
                event,
                context,
                query: {
                    ...(request.params || {}),
                    ...(request.data ? { data: request.data || {} } : {}),
                },
            };
            try {
                if ((config['statics'] || {})[request.uri]) {
                    const staticInfos = config['statics'][request.uri];
                    return helpers.staticFile({
                        root: config.root,
                        ...staticInfos,
                        headers: {
                            ...(((_a = ctx.config) === null || _a === void 0 ? void 0 : _a.defaultStaticsHeaders) || {}),
                            ...(staticInfos.headers || {})
                        }
                    });
                }
            }
            catch (e1) {
                return helpers.httpError({ error: e1, phase: 'static' });
            }
            try {
                return await (customProcessHandler || defaultProcessHandler_1.default)(ctx);
            }
            catch (e) {
                return helpers.httpError({
                    error: e,
                    phase: 'process',
                    headers: {
                        ...(((_b = request.route) === null || _b === void 0 ? void 0 : _b.name) ? { 'X-Error-Route-Name': request.route.name } : {}),
                        ...(((_c = request.route) === null || _c === void 0 ? void 0 : _c.type) ? { 'X-Error-Route-Type': request.route.type } : {}),
                    }
                });
            }
        }
        catch (e0) {
            return helpers.httpError({ error: e0, phase: 'init' });
        }
    };
}
exports.createRouterHandler = createRouterHandler;
exports.default = createRouterHandler;
