// @ohoareau/react-document-renderer/lib/blocks/ListBlock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const Block_1 = __importDefault(require("../Block"));
const renderer_1 = require("@react-pdf/renderer");
const hocs_1 = require("../hocs");
exports.ListBlock = hocs_1.pdfComponent({
    list: {},
    list_item: {
        flexDirection: 'row',
    },
    list_item_bullet_point: {},
    list_item_content: {
        flex: 1,
    },
}, ({ s = () => { }, block }) => (react_1.default.createElement(renderer_1.View, { style: s('list') }, (block.items || []).map((item, i) => (react_1.default.createElement(renderer_1.View, { key: i, style: s('list_item') },
    react_1.default.createElement(renderer_1.Text, { style: s('list_item_bullet_point') }, "\u2022"),
    react_1.default.createElement(renderer_1.View, { style: s('list_item_content') },
        react_1.default.createElement(Block_1.default, { id: item.id, block: { pseudo: true, ...item } }))))))));
exports.default = exports.ListBlock;
