// @ohoareau/lambda-api-document/lib/actions/generate.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const lambda_utils_1 = require("@ohoareau/lambda-utils");
const availableSources = __importStar(require("../sources"));
const docx_1 = require("docx");
async function render(def) {
    const doc = new docx_1.Document();
    doc.addSection({
        properties: {},
        children: [
            new docx_1.Paragraph({
                children: [
                    new docx_1.TextRun("Hello World"),
                    new docx_1.TextRun({
                        text: "Foo Bar",
                        bold: true,
                    }),
                    new docx_1.TextRun({
                        text: "\tGithub is the best",
                        bold: true,
                    }),
                ],
            }),
        ],
    });
    return docx_1.Packer.toBuffer(doc);
}
async function generate(ctx) {
    const sources = { ...availableSources, ...(ctx.config.sources || {}) };
    const sourceName = ctx.query.source || 'default';
    const source = sources[sourceName];
    if (!source)
        return ctx.helpers.httpNotFound();
    const sourced = (await source(ctx)) || {};
    if (!sourced || !sourced.definition)
        return ctx.helpers.httpNotFound();
    try {
        return ctx.helpers.buffer({ contentType: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document', ...sourced, buffer: await render(sourced.definition) });
    }
    catch (e) {
        throw new lambda_utils_1.UnderlyingExecutionError(e, 'generate', { sourceName });
    }
}
exports.generate = generate;
exports.default = generate;
