// @ohoareau/lambda-api-image/lib/utils/buildOrder.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const lambda_utils_1 = require("@ohoareau/lambda-utils");
const availableSourceTypes = __importStar(require("../sources"));
const imageman_1 = require("@ohoareau/imageman");
const modifiers = __importStar(require("../modifiers"));
async function buildOrder(ctx) {
    var _a, _b, _c;
    const modifierNames = [
        'preset', 'size', 'trim', 'flip', 'format', 'radius', 'filter',
        'rotation', 'theme', 'color', 'quality',
    ];
    let { input = undefined, operations = [], options = undefined, format } = await modifierNames.reduce(async (acc, modifier) => {
        acc = await acc;
        if (!modifiers[modifier])
            return acc;
        return (await modifiers[modifier](acc, ctx.request, ctx.config)) || acc;
    }, Promise.resolve({ input: undefined, operations: [], options: {} }));
    const sources = { ...availableSourceTypes, ...(ctx.config.sources || {}) };
    const source = sources[ctx.query.source || 'default'];
    if (!source)
        throw new lambda_utils_1.ResourceNotFoundError(ctx.request);
    const sourced = (await source(ctx)) || {};
    sourced.input && (input = sourced.input);
    sourced.operations && (operations = sourced.operations);
    sourced.options && (options = sourced.options);
    sourced.format && (format = sourced.format);
    let formatFromFile = undefined;
    let contentTypeFromFile = undefined;
    if ((_a = ctx.query) === null || _a === void 0 ? void 0 : _a.file) {
        formatFromFile = imageman_1.detectFormatFromFileName((_b = ctx.query) === null || _b === void 0 ? void 0 : _b.file);
        contentTypeFromFile = lambda_utils_1.detectContentTypeFromFileName((_c = ctx.query) === null || _c === void 0 ? void 0 : _c.file);
    }
    const opts = {
        ...(options || {}),
        ...((contentTypeFromFile || (options === null || options === void 0 ? void 0 : options.contentType))
            ? { contentType: contentTypeFromFile || (options === null || options === void 0 ? void 0 : options.contentType) }
            : {}),
    };
    return {
        input,
        format: formatFromFile || format,
        operations: [
            ...operations,
            ...(operations || []),
        ],
        options: opts,
        sourceTypes: {
            ...(ctx.config.sourceTypes || {}),
        },
        targetTypes: {
            ...(ctx.config.targetTypes || {}),
        },
        output: 'buffer',
    };
}
exports.buildOrder = buildOrder;
exports.default = buildOrder;
