// @ohoareau/errors/lib/InvokeError.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AbstractError_1 = __importDefault(require("./AbstractError"));
class InvokeError extends AbstractError_1.default {
    constructor(type, operation, dsn, requestPayload, responsePayload) {
        super(`Invoking operation '${operation}' on service '${type}' raise an error when invoking '${dsn}': ${responsePayload.errorMessage} (errorType: ${responsePayload.errorType})`, 500, 'invoke', {}, { type, operation, dsn });
        this.type = type;
        this.operation = operation;
        this.dsn = dsn;
        this.requestPayload = requestPayload;
        this.responsePayload = responsePayload;
    }
    getType() {
        return this.type;
    }
    getOperation() {
        return this.operation;
    }
    getDsn() {
        return this.dsn;
    }
    getRequestPayload() {
        return this.requestPayload;
    }
    getResponsePayload() {
        return this.responsePayload;
    }
}
exports.InvokeError = InvokeError;
exports.default = InvokeError;
