// @ohoareau/errors/lib/AbstractError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class AbstractError extends Error {
    constructor(message, code, errorType = 'error', data = {}, errorInfo = {}) {
        super(message);
        this.code = 500;
        this.data = {};
        this.errorType = 'error';
        this.errorInfo = {};
        this.shortMessage = 'Error';
        this.setShortMessage(message.split(/\n/)[0].trim());
        this.setCode(code);
        this.setErrorType(errorType);
        this.setData(data);
        this.setErrorInfo(errorInfo);
    }
    getCode() {
        return this.code;
    }
    getMessage() {
        return this.message;
    }
    getShortMessage() {
        return this.shortMessage;
    }
    getErrorType() {
        return this.errorType;
    }
    getData() {
        return this.data;
    }
    getErrorInfo() {
        return this.errorInfo;
    }
    serialize() {
        return {
            errorType: this.getErrorType(),
            message: this.getShortMessage(),
            code: this.getCode(),
            data: this.getData(),
            errorInfo: this.getErrorInfo(),
        };
    }
    setCode(code) {
        this.code = code;
    }
    setShortMessage(shortMessage) {
        this.shortMessage = shortMessage;
    }
    setErrorType(errorType) {
        this.errorType = errorType;
    }
    setData(data) {
        this.data = data;
    }
    setErrorInfo(errorInfo) {
        this.errorInfo = errorInfo;
    }
}
exports.AbstractError = AbstractError;
exports.default = AbstractError;
